src/neurodsp/raw_metrics.py

```diff
@@ -0,0 +1,126 @@
+from neurodsp.voltage import destripe, destripe_lfp, detect_bad_channels
+from neurodsp.utils import rms
+import spikeglx
+import pandas as pd
+import scipy
+from tqdm import trange
+import numpy as np
+
+
+def raw_data_features(ap_cbin, lf_cbin, t_start, t_end):
+    """
+    Create a QC table with the following features for each channel for
+    each snippets:
+
+    - RMS of raw, butterworth, destriped, and residual (butterworth - destriped)
+    for both bands (returned in uV).
+
+    - DC offset of raw data for both bands (returned in uV).
+
+    - Channel labels estimated from both AP and LF bands separately.
+
+    - Channel detection features (LF and HF cross-correlation and HF PSD)
+    estimated from both AP and LF bands separately.
+
+    AP band data is filtered with a 300Hz high-pass butterworth for
+    the BTR and DST features. LF band is filtered with a 2-200Hz band-pass
+    butterworth. Snippets of 12x the length at the same starting times are
+    taken from LF to account for different sampling rates.
+
+    Returns a Pandas Multiindex dataframe with snippet and channel ids.
+
+    :param ap_cbin: Path to AP band cbin.
+    :param lf_cbin: Path ot LF band cbin.
+    :param t_start: Starting times of snippets in seconds.
+    :param t_end: ending times of snippets in seconds.
+    """
+    assert len(t_start) == len(t_end), "mismatched snippet start and end"
+    num_snippets = len(t_start)
+
+    sr_ap = spikeglx.Reader(ap_cbin)
+    sr_lf = spikeglx.Reader(lf_cbin)
+
+    filter_ap = scipy.signal.butter(
+        N=3, Wn=300 / sr_ap.fs * 2, btype="highpass", output="sos"
+    )
+    filter_lf = scipy.signal.butter(
+        N=3, Wn=[2 / sr_lf.fs * 2, 200 / sr_ap.fs * 2], btype="bandpass", output="sos"
+    )
+
+    dfs = {}
+
+    print(f"Extracting features over {num_snippets} snippets")
+    for i in trange(num_snippets):
+        t0 = t_start[i]
+        t1 = min(t_end[i], sr_ap.rl)
+        df = compute_raw_features_snippet(sr_ap, sr_lf, t0, t1, filter_ap, filter_lf)
+        dfs[i] = df
+
+    out_df = pd.concat(dfs)
+    out_df.index.rename(["snippet_id", "channel_id"], inplace=True)
+
+    return out_df
+
+
+def compute_raw_features_snippet(sr_ap, sr_lf, t0, t1, filter_ap=None, filter_lf=None):
+    """
+    Compute raw data feature table on one snippet of AP / LF data.
```

Comment:
Just add that the needed input is a scipy sos filter

Reply:
fixd